File: 0841-keys-and-rooms/0841-keys-and-rooms.cpp
class Solution {
public:

    void dfs(int i, set<int>& keys, vector<vector<int>>& rooms, vector<int>& vis){
        vis[i]=1;
        for(auto it: rooms[i]){
            if(vis[it]==0){
                keys.insert(it);
                dfs(it, keys, rooms, vis);
            }
        }
    }

    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        set<int> keys;
        keys.insert(0);
        vector<int> vis(n, 0);

        for(int i=0; i<n; i++){
            if(keys.find(i)!=keys.end() && vis[i]==0){
                dfs(i, keys, rooms, vis);
            }
        }
        for(int i=0; i<n; i++){
            if(vis[i]==0){
                return false;
            }
        }
        return true;
    }
};

File: 0841-keys-and-rooms/0841-keys-and-rooms.py
class Solution:
    
    def dfs(self, i, keys, rooms: List[List[int]], vis: List[int]):
        vis[i]=1
        for j in rooms[i]:
            if vis[j]==0:
                keys.add(j)
                self.dfs(j, keys, rooms, vis)


    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        keys = {0}
        # keys.add(0)
        vis = [0]*n

        for i in range(n):
            if i in keys and vis[i]==0:
                self.dfs(i, keys, rooms, vis)
        
        for i in range(n):
            if vis[i]==0:
                return False
        
        return True